### PR TITLE
[MPLUGIN-508] Upgrade to Maven 4.0.0-beta-3

### PR DESCRIPTION
**Pull request by gnodet**
- **[MPLUGIN-508] Upgrade to 4.0.0-beta-3**

### CODE REVIEW

**Comment by michael-o** [comment]
Please hold this one off until #287 and 3.13.1 has been cut. I want to clean up first before we cut 3.x.

File: maven-plugin-plugin/src/it/v4api/src/main/java/org/apache/maven/its/v4api/FirstMojo.java
Comment (michael-o) [inline maven-plugin-plugin/src/it/v4api/src/main/java/org/apache/maven/its/v4api/FirstMojo.java:46]: Is there a reason why other args to `@Mojo` are gone?

File: maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/DescriptorGeneratorMojo.java
Comment (michael-o) [inline maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/DescriptorGeneratorMojo.java:97]: `classesOutputDirectory`?

File: maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/DescriptorGeneratorMojo.java
Comment (michael-o) [inline maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/DescriptorGeneratorMojo.java:449]: `System.lineSeparator()`?

**Review by michael-o — CHANGES_REQUESTED** [review]

File: maven-plugin-plugin/src/it/v4api/src/main/java/org/apache/maven/its/v4api/FirstMojo.java
Comment (gnodet) [reply maven-plugin-plugin/src/it/v4api/src/main/java/org/apache/maven/its/v4api/FirstMojo.java:46]: Maven 4 mojos have to be thread safe.
In maven 3, the `requiresDependencyResolution` and `requiresDependencyCollection` are currently used to collect or resolve the project dependencies in a given scope.  Those artifacts are then made available by Maven by `MavenProject.getArtifacts()`.  This brings several problems: for a single project, two mojos can not be run concurrently, and as we've seen, some mojos even require runtime + test scope resolution, and that cannot be provided.   In 4.x, mojos will have to use the API to retrieve collect/resolve the dependencies.  Hopefully the result can be cached in the session.  So the latest 4.x release did remove those two bits.  However, not having this information sounds wrong, as this would be needed in at least two scenarii:
 * it would be nice to be able to eagerly download dependencies at the beginning of a given build, and we'd need to know which scopes are needed for a given project, this could be done by analysing the build plan
 * if we experiment a bit more with the [fully concurrent builder](https://github.com/apache/maven/pull/1429), we need this information in order to know which goals depend on which projects in the reactor, so that the goals can be scheduled correctly
So I think we'll need to add those back somehow, but they would be informational.  Another possibility would be to allow injecting the resolution/collection result in the mojo.

File: maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/DescriptorGeneratorMojo.java
Comment (gnodet) [reply maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/DescriptorGeneratorMojo.java]: Fixed

File: maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/DescriptorGeneratorMojo.java
Comment (gnodet) [reply maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/DescriptorGeneratorMojo.java]: Fixed

File: maven-plugin-plugin/src/it/v4api/src/main/java/org/apache/maven/its/v4api/FirstMojo.java
Comment (gnodet) [reply maven-plugin-plugin/src/it/v4api/src/main/java/org/apache/maven/its/v4api/FirstMojo.java:46]: I raised https://issues.apache.org/jira/browse/MNG-8134

Comment (michael-o) [reply maven-plugin-plugin/src/it/v4api/src/main/java/org/apache/maven/its/v4api/FirstMojo.java:46]: > Maven 4 mojos have to be thread safe, so that annotation is gone.

You cannot enforce that right? Only by contract...

File: maven-plugin-plugin/src/it/skip/src/main/java/org/apache/maven/plugin/coreit/FirstMojo.java
Comment (michael-o) [inline maven-plugin-plugin/src/it/skip/src/main/java/org/apache/maven/plugin/coreit/FirstMojo.java:29]: This is now completely different, how?

File: maven-plugin-plugin/src/it/v4api/invoker.properties
Comment (michael-o) [inline maven-plugin-plugin/src/it/v4api/invoker.properties:18]: Confusing if the entire project require Java 17, why here especially?

File: maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/DescriptorGeneratorMojo.java
Comment (michael-o) [inline maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/DescriptorGeneratorMojo.java:59]: Is this necessary?

**Review by michael-o — CHANGES_REQUESTED** [review]

File: maven-plugin-plugin/src/it/v4api/src/main/java/org/apache/maven/its/v4api/FirstMojo.java
Comment (gnodet) [reply maven-plugin-plugin/src/it/v4api/src/main/java/org/apache/maven/its/v4api/FirstMojo.java:46]: > > Maven 4 mojos have to be thread safe, so that annotation is gone.
> 
> You cannot enforce that, right? Only by contract...

Right.

File: maven-plugin-plugin/src/it/v4api/invoker.properties
Comment (gnodet) [reply maven-plugin-plugin/src/it/v4api/invoker.properties:18]: The entire project is JDK 8.  The mojos run on Maven 3.x + JDK 8 if you want to build Maven 3.x plugins.  If you build a Maven 4.x plugin, you'll need JDK 17.  That's the case for this IT.  Other ITs are unchanged.

The migration to Maven 4 only should imho happen at the same time as m-plugin-p is migrated to be a Maven 4 plugin.

File: maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/DescriptorGeneratorMojo.java
Comment (gnodet) [reply maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/DescriptorGeneratorMojo.java:59]: The imports are used to generate the factories for Maven 4.x mojos.  They are not using plexus, as they are booted by the new DI framework in Maven.  But this framework recognises beans annotated with `@org.apache.maven.api.di.Named`, which is not the case for mojos.  To avoid forcing the user to add this annotation, a factory (annotated with the needed annotation) is generated by the plugin.  And to avoid having to do another compile phase, the byte code is directly generated instead of generating source code.  Those new imports are needed for this feature.

File: maven-plugin-report-plugin/src/it/mplugin-319_report-since/src/main/java/org/MyMojo.java
Comment (gnodet) [reply maven-plugin-report-plugin/src/it/mplugin-319_report-since/src/main/java/org/MyMojo.java:33]: Why ? if not specified, the plugin won't have a default phase, but that's not a problem here.  We don't even use this plugin, we're only building it.

File: maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/DescriptorGeneratorMojo.java
Comment (michael-o) [reply maven-plugin-plugin/src/main/java/org/apache/maven/plugin/plugin/DescriptorGeneratorMojo.java:59]: Accepted.

File: maven-plugin-report-plugin/src/it/mplugin-319_report-since/src/main/java/org/MyMojo.java
Comment (michael-o) [reply maven-plugin-report-plugin/src/it/mplugin-319_report-since/src/main/java/org/MyMojo.java:33]: The previous config had a lot of stuff:
```
 * @goal noop
 * @phase process-sources
 * @requiresDependencyResolution test
 * @requiresDirectInvocation true
 * @requiresOnline
 * @inheritByDefault false
 * @execute phase="compile"
 * @aggregator
 ```

Plain `@Mojo` isn't an equivalent, is it?

File: maven-plugin-report-plugin/src/main/java/org/apache/maven/plugins/plugin/descriptor/EnhancedPluginDescriptorBuilder.java
Comment (michael-o) [reply maven-plugin-report-plugin/src/main/java/org/apache/maven/plugins/plugin/descriptor/EnhancedPluginDescriptorBuilder.java:45]: Alright, let's keep the wording as-is.

File: maven-plugin-report-plugin/src/it/mplugin-319_report-since/src/main/java/org/MyMojo.java
Comment (gnodet) [reply maven-plugin-report-plugin/src/it/mplugin-319_report-since/src/main/java/org/MyMojo.java:33]: No, but again, those are not tested.  I think the one creating the original test just copy/pasted an existing one.

File: maven-plugin-plugin/src/it/skip/src/main/java/org/apache/maven/plugin/coreit/FirstMojo.java
Comment (gnodet) [reply maven-plugin-plugin/src/it/skip/src/main/java/org/apache/maven/plugin/coreit/FirstMojo.java:29]: Those mojos are just built and used to verify a few specific things depending on each test.  The annotations are usually not used or tested.

**Review by michael-o — APPROVED** [review]
I have no further objections, though two others should have a look as well.

**Pull request merged**
